### PR TITLE
SVA-183 | Shared projects & events container

### PR DESCRIPTION
**Pull request by joefhall**
- [Description](#description)
  - [Type of change](#type-of-change)
- [Checklist](#checklist)

# Description

- Projects container, Projects Search Results container and Events container were all doing very similar jobs, as were some of the components to show those lists of things
- Created a shared List container and shared List component that can be used by projects, events and other kinds of things in the future if needed, reducing our code base
- Set up Redux store for projects so projects data can be shared across components, also avoids previous situation where projects data was being loaded twice (once in main Projects container, again in Projects Search container)
- Made a couple of small tweaks (e.g. option to clear search results), otherwise functionality should work as it does on the existing app

Fixes SVA 183

## Type of change

Please delete options that are not relevant
- [x] New Feature (non-breaking change which adds functionality)

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by DougieWougie — CHANGES_REQUESTED** [review]
I ran @rc-1988 new elint command, which runs but throws fair few linting errors! Mostly `Missing @file` though.

[linting_errors.md](https://github.com/Scottish-Tech-Army/Volunteer-app/files/9959653/linting_errors.md)

**Comment by joefhall** [comment]
Thanks @DougieWougie -- apologies, I wasn't ready for review yet but hadn't marked this as draft (I was sneakily publishing the PR to see what ES Lint errors I got back!)

I'll give you a shout when it's ready to look at properly, thanks for picking it up so quickly

**Comment by DougieWougie** [comment]
> Thanks @DougieWougie -- apologies, I wasn't ready for review yet but hadn't marked this as draft (I was sneakily publishing the PR to see what ES Lint errors I got back!)
> 
> I'll give you a shout when it's ready to look at properly, thanks for picking it up so quickly

Didn't notice it was a draft, thought it followed on from the other Linting issues, my bad!

Running elint from the terminal requires me to go into WSL and run it from Ubuntu, so I get where you're coming from. I can't find a PowerShell equivalent to xargs.

**Comment by joefhall** [comment]
@DougieWougie no, my bad 🙂   All good for review now if you have a chance to look -- thanks

File: app/src/Components/List.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            )
+          }}
+        />
+      )
+      break
+  }
+
+  return data.length ? (
+    flatList
+  ) : (
```
Comment (jppferguson) [inline app/src/Components/List.tsx:101]: You could probably move this handling of empty data  above your switch statement, then you can just return early and you wouldn't need to assign the `flatlist` variable. eg. something like
```tsx
if(!data.length) {
	return (
		<>
			<NoneFound>
				. . .
			</NoneFound>
			{mode === 'search' && (
				. . .
			)}
		</>
        )
}

return (
  switch (type) {
    case ListType.Events:
	return <FlatList data={data as Events} />
    case ListType.Projects:
	return <FlatList data={data as Projects} />
  }
)

```

**Review by jppferguson — COMMENTED** [review]
Nice work @joefhall – this tidies things up a bit! Just a small suggestion on making the list component slightly more readable

File: app/src/Components/List.tsx (hunk 1 above)
Comment (joefhall) [reply app/src/Components/List.tsx:117]: Nice one, thanks @jppferguson -- will do!

**Review by DougieWougie — APPROVED** [review]
Thanks @joefhall !

**Review by CatMac228 — APPROVED** [review]
No issues on IOS :)

**Comment by joefhall** [comment]
Thanks a lot @DougieWougie @jppferguson @CatMac228 ! 🙌

**Pull request merged**